src/numerics/petsc_matrix.C

```diff
@@ -1201,6 +1201,53 @@ void PetscMatrix<T>::add (const T a_in, const SparseMatrix<T> & X_in)
 }
 
 
+template <typename T>
+void PetscMatrix<T>::mat_mult (SparseMatrix<T> & X_in, SparseMatrix<T> & Y_out)
+{
+  libmesh_assert (this->initialized());
+
+   // sanity check.
+   libmesh_assert_equal_to (this->n(), X_in.m());
+   libmesh_assert_equal_to (this->m(), Y_out.m());
+   libmesh_assert_equal_to (X_in.n(), Y_out.n());
+
+   const PetscMatrix<T> * X = cast_ptr<const PetscMatrix<T> *> (&X_in);
+   PetscMatrix<T> * Y = cast_ptr<PetscMatrix<T> *> (&Y_out);
+
+   libmesh_assert (X);
+   libmesh_assert (Y);
+
+   PetscErrorCode ierr=0;
+
+   // the matrix from which we copy the values has to be assembled/closed
+   libmesh_assert(X->closed());
+
+   semiparallel_only();
+
+   ierr = MatMatMult(_mat, X->_mat, MAT_INITIAL_MATRIX, PETSC_DEFAULT, &Y->_mat);
+   LIBMESH_CHKERR(ierr);
+
+}
+
+template <typename T>
+void PetscMatrix<T>::add_sparse_matrix (const SparseMatrix<T> & spm,
+                        const std::vector<numeric_index_type> & rows,
+                        const std::vector<numeric_index_type> & cols,
+                        const T scalar)
+{
+  libmesh_assert_equal_to(spm.m(), rows.size());
+  libmesh_assert_equal_to(spm.n(), cols.size());
+
+  for(unsigned int i = 0; i<rows.size(); ++i)
+  {
+    std::vector<numeric_index_type> indices;
+    std::vector<T> values;
+    spm.get_row(i, indices, values);
```

Comment:
This is not quite right in parallel. `i ` won't start from zero for nonzero MPI ranks.

Reply:
> This is not quite right in parallel. `i ` won't start from zero for nonzero MPI ranks.

That is a good point. So what to do here?

Reply:
Could you push up your moose changes? We should start from the caller

Reply:
`const std::map<numeric_index_type> & row_ltog,`

and then

```
for (auto ltog: row_ltog)
   spm.get_row(ltog.first, indices, values);
```

Reply:
> Could you push up your moose changes? We should start from the caller

Please look at an initial implementation for condensing out the LMs [here](https://github.com/idaholab/moose/pull/15565)